File: Shared/Helpers/GameSituationHelper.cs
namespace FourPlayWebApp.Shared.Helpers;

public static class GameSituationHelper
{
    /// <summary>
    /// Analyzes the current game state for Home/Away spreads and Over/Under lines.
    /// Only returns text when a team is “behind” on a bet; if already covered, leaves empty.
    /// </summary>
    /// <param name="homeTeam"></param>
    /// <param name="awayTeam"></param>
    /// <param name="homeScore"></param>
    /// <param name="awayScore"></param>
    /// <param name="homeSpread"></param>
    /// <param name="awaySpread"></param>
    /// <param name="overLine"></param>
    /// <param name="underLine"></param>
    /// <returns></returns>
    public static string? GetSituations(
        string homeTeam,
        string awayTeam,
        long homeScore,
        long awayScore,
        double? homeSpread = null,
        double? awaySpread = null,
        double? overLine = null,
        double? underLine = null) {
        string output = "";
        var total = homeScore + awayScore;

        // --- Home spread ---
        if (homeSpread.HasValue)
        {
            double coverMargin = (homeScore + homeSpread.Value) - awayScore;
            if (coverMargin < 0)
            {
                int needed = (int)Math.Ceiling(Math.Abs(coverMargin));
                string scoreText = PointsToText(needed);
                output = $"{homeTeam} needs {scoreText} to cover the spread ({homeSpread}).";
            }
        }

        // --- Away spread ---
        if (awaySpread.HasValue)
        {
            double coverMargin = (awayScore + awaySpread.Value) - homeScore;
            if (coverMargin < 0)
            {
                int needed = (int)Math.Ceiling(Math.Abs(coverMargin));
                string scoreText = PointsToText(needed);
                output = $"{awayTeam} needs {scoreText} to cover the spread ({awaySpread}).";
            }
        }

        // --- Over ---
        if (total < overLine)
        {
            var needed = overLine.Value - total;
            output = $"The game needs {needed} more points to hit the Over ({overLine}).";
        }
        // --- Under ---
        if (!(total > underLine)) return output;
        var extra = total - underLine.Value;
        output = $"The game has exceeded the Under by {extra} points ({underLine}).";

        return output;
    }

    // Simple football scoring text helper
    private static string PointsToText(int points)
    {
        return points switch {
            <= 3 => "one FG",
            <= 7 => "one score",
            <= 14 => "two scores",
            <= 21 => "three scores",
            _ => $"{(points / 7) + 1} scores"
        };
    }
}

File: Client.React/src/utils/gameSituationHelper.ts
export function getSituations(
  homeTeam: string,
  awayTeam: string,
  homeScore: number,
  awayScore: number,
  homeSpread?: number | null,
  awaySpread?: number | null,
  overLine?: number | null,
  underLine?: number | null
): string | null {
  let output = '';
  const total = homeScore + awayScore;

  if (homeSpread !== null && homeSpread !== undefined) {
    const coverMargin = homeScore + homeSpread - awayScore;
    if (coverMargin < 0) {
      const needed = Math.ceil(Math.abs(coverMargin));
      output = `${homeTeam} needs ${pointsToText(needed)} to cover the spread (${homeSpread}).`;
    }
  }

  if (awaySpread !== null && awaySpread !== undefined) {
    const coverMargin = awayScore + awaySpread - homeScore;
    if (coverMargin < 0) {
      const needed = Math.ceil(Math.abs(coverMargin));
      output = `${awayTeam} needs ${pointsToText(needed)} to cover the spread (${awaySpread}).`;
    }
  }

  if (overLine !== null && overLine !== undefined && total < overLine) {
    const needed = overLine - total;
    output = `The game needs ${needed} more points to hit the Over (${overLine}).`;
  }

  if (underLine !== null && underLine !== undefined && total > underLine) {
    const extra = total - underLine;
    output = `The game has exceeded the Under by ${extra} points (${underLine}).`;
  }

  return output || null;
}

function pointsToText(points: number) {
  if (points <= 3) return 'one FG';
  if (points <= 7) return 'one score';
  if (points <= 14) return 'two scores';
  if (points <= 21) return 'three scores';
  return `${Math.floor(points / 7) + 1} scores`;
}
